src/naive/time/u31.rs

```diff
@@ -0,0 +1,287 @@
+use core::{cmp, fmt, hash, mem};
+
+/// A 31-bit unsigned integer
+#[allow(unreachable_pub)] // public through `rkyv::Archive<NaiveTime>`
+#[repr(transparent)]
+#[derive(Copy, Clone)]
+pub struct U31(Buf);
+
+impl U31 {
+    pub(crate) const ZERO: U31 = { U31(Buf { align: [], a: 0, b: 0, c: 0, d: MaxByte::V0 }) };
+
+    pub(crate) const SEC_MINUS_1: Self = {
+        // [int(x, 16) for x in reversed(re.findall('..', f'{999_999_999:x}'))]
+        U31(Buf { align: [], a: 255, b: 201, c: 154, d: MaxByte::V59 })
+    };
+
+    /// SAFETY: the caller has to ensure that the value is in range
+    #[inline]
+    pub(crate) unsafe fn new_unchecked(value: u32) -> Self {
+        mem::transmute(value)
+    }
+
+    #[inline]
+    pub(crate) fn get(self) -> u32 {
+        unsafe { mem::transmute(self) }
+    }
+}
+
+impl fmt::Debug for U31 {
+    #[inline]
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        self.get().fmt(f)
+    }
+}
+
+impl fmt::Display for U31 {
+    #[inline]
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        self.get().fmt(f)
+    }
+}
+
+impl cmp::PartialEq for U31 {
+    #[inline]
+    fn eq(&self, other: &Self) -> bool {
+        self.get() == other.get()
+    }
+}
+
+impl cmp::PartialOrd for U31 {
+    #[inline]
+    fn partial_cmp(&self, other: &Self) -> Option<cmp::Ordering> {
+        self.get().partial_cmp(&other.get())
+    }
+}
+
+impl cmp::Eq for U31 {}
+
+impl cmp::Ord for U31 {
+    #[inline]
+    fn cmp(&self, other: &Self) -> cmp::Ordering {
+        self.get().cmp(&other.get())
+    }
+}
+
+impl hash::Hash for U31 {
+    fn hash<H: hash::Hasher>(&self, state: &mut H) {
+        self.get().hash(state);
+    }
+}
+
+impl Default for U31 {
+    #[inline]
+    fn default() -> Self {
+        Self::ZERO
+    }
+}
+
+#[repr(C)]
+#[cfg(target_endian = "little")]
+#[derive(Copy, Clone)]
+struct Buf {
+    align: [u32; 0],
```

Comment:
Is this trick documented anywhere? I'd like this to be commented with a (more or less) authoritative citation on how this behaves.

Reply:
> Slices have the same layout as the section of the array they slice. -- [Type layout # Slice layout](https://doc.rust-lang.org/reference/type-layout.html#slice-layout)

This is the same for 0 sized slices.

Reply:
For information, you can also use [`#[repr(align(4))]`](https://doc.rust-lang.org/reference/type-layout.html#representations) to do the same thing (from [RFC 1358](https://github.com/rust-lang/rfcs/blob/master/text/1358-repr-align.md)).

Reply:
For some platforms the alignment of u32 might be 16bits. msp430-none-elf is a 16bit, [tier 3](https://doc.rust-lang.org/nightly/rustc/platform-support.html) platform.